polkadot/xcm/pallet-xcm/src/precompiles.rs

```diff
@@ -0,0 +1,165 @@
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) =>
+				frame_system::RawOrigin::Signed(account_id.clone()).into(),
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let _ = env.charge(<Runtime as Config>::WeightInfo::send())?;
+
+				let final_destination = VersionedLocation::decode_all(&mut &destination[..])
+					.map_err(|error| {
+						error!(target: "xcm::precompiles", ?error, "XCM send failed: Invalid destination format");
+						Error::Revert("XCM send failed: Invalid destination format".into())
+					})?;
+
+				let final_message = VersionedXcm::<()>::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					error!(target: "xcm::precompiles", ?error, "XCM send failed: Invalid message format");
+					Error::Revert("XCM send failed: Invalid message format".into())
+				})?;
+
+				crate::Pallet::<Runtime>::send(
+					frame_origin,
+					final_destination.into(),
+					final_message.into(),
+				)
+				.map(|message_id| message_id.encode())
+				.map_err(|error| {
+					error!(
+						target: "xcm::precompiles",
+						?error,
+						"XCM send failed: destination or message format may be incompatible"
+
+					);
+					Error::Revert(
+						"XCM send failed: destination or message format may be incompatible".into(),
+					)
+				})
+			},
+			IXcmCalls::xcmExecute(IXcm::xcmExecuteCall { message, weight }) => {
+				let weight = Weight::from_parts(weight.refTime, weight.proofSize);
+				let charged_amount = env.charge(weight)?;
+
+				let final_message = VersionedXcm::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					error!(target: "xcm::precompiles", ?error, "XCM execute failed: Invalid message format");
+					Error::Revert("Invalid message format".into())
+				})?;
+
+				crate::Pallet::<Runtime>::execute(frame_origin, final_message.into(), weight)
+					.map(|post_dispatch_info| {
+						// Adjust charged amount to account for the actual weight consumed.
+						if let Some(actual_weight) = post_dispatch_info.actual_weight {
+							env.gas_meter_mut().adjust_gas(
+								charged_amount.clone(),
+								RuntimeCosts::Precompile(actual_weight),
+							);
+						}
+						post_dispatch_info.encode()
+					})
+					.map_err(|error| {
+						error!(
+							target: "xcm::precompiles",
+							?error,
+							"XCM execute failed: message may be invalid or execution constraints not satisfied"
+						);
+						// Adjust charged amount to account for the actual weight consumed.
+						if let Some(actual_weight) = error.post_info.actual_weight {
+							env.gas_meter_mut().adjust_gas(
+								charged_amount,
+								RuntimeCosts::Precompile(actual_weight),
+							);
```

Comment:
maybe we can add `adjust_gas` to the Precompile trait
so that we can just write
```rust
 env.adjust_gas(charged_amount, actual_weight);
```